src/Service/ForecastReportService.php

```diff
@@ -0,0 +1,192 @@
+
+                // Check if worklog is billed
+                $isWorklogBilled = $worklog->isBilled();
+
+                // Tally up total project hours based on whether the worklog is billed
+                $currentProject->invoiced += $worklogTime;
+                if ($isWorklogBilled) {
+                    $currentProject->invoicedAndRecorded += $worklogTime;
+                }
+
+                // Get issue details from worklog
+                $issueId = $worklog->getIssue()->getProjectTrackerKey();
+                $issueLink = $worklog->getIssue()->getLinkToIssue();
+                $issueTag = $worklog->getIssue()->getEpicName() ?: '[no tag]';
+
+                // Add issue in the project if it does not exist
+                if (!isset($currentProject->issues[$issueTag])) {
+                    $currentProject->issues[$issueTag] = new ForecastReportIssueData($issueTag);
+                    $currentProject->issues[$issueTag]->issueId = $issueId;
+                    $currentProject->issues[$issueTag]->issueLink = $issueLink;
+                }
+
+                // Get current issue from project
+                $currentIssue = $currentProject->issues[$issueTag];
+
+                // Add up the invoiced hours to the current issue
+                $currentIssue->invoiced += $worklogTime;
+                if ($isWorklogBilled) {
+                    $currentIssue->invoicedAndRecorded += $worklogTime;
+                }
+
+                // Get version details from issue
+                $issueVersions = $worklog->getIssue()->getVersions();
+                $issueVersion = count($issueVersions) > 0 ? implode(', ', array_map(function ($version) { return $version->getName(); }, $issueVersions->toArray())) : '[no version]';
+
+                $issueVersionIdentifier = $issueTag.$issueVersion;
+
+                // Add version entry in the issue if it does not exist
+                if (!isset($currentIssue->versions[$issueVersion])) {
+                    $currentIssue->versions[$issueVersion] = new ForecastReportIssueVersionData($issueVersion);
+                    $currentIssue->versions[$issueVersion]->issueVersionIdentifier = $issueVersionIdentifier;
+                }
+
+                // Get the current version from issue
+                $currentVersion = $currentIssue->versions[$issueVersion];
+
+                // Add up invoiced hours in current version
+                $currentVersion->invoiced += $worklogTime;
+
+                // If worklog is billed, add it to the recorded hours as well
+                if ($isWorklogBilled) {
+                    $currentVersion->invoicedAndRecorded += $worklogTime;
+                }
+
+                // Get worklog details
+                $worklogId = $worklog->getId();
+                $workerEmail = $worklog->getWorker();
+                $worker = $this->workerRepository->findOneBy(['email' => $workerEmail]);
```

Comment:
Extract alle worker email -> name mappings out of the loops to avoid unnecessary db lookups.